File: sapl-vaadin-editor/src/main/resources/META-INF/resources/frontend/json-editor-lsp.js
/*
 * Copyright (C) 2017-2025 Dominic Heutelbeck (dominic@heutelbeck.com)
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * JSON Editor using CodeMirror 6 with syntax highlighting, linting, and merge view.
 */
import { LitElement, html, css } from 'lit';
import { EditorView, basicSetup } from 'codemirror';
import { EditorState, Compartment } from '@codemirror/state';
import { keymap } from '@codemirror/view';
import { indentWithTab } from '@codemirror/commands';
import { oneDark } from '@codemirror/theme-one-dark';
import { json, jsonParseLinter } from '@codemirror/lang-json';
import { linter, lintGutter } from '@codemirror/lint';
import { bracketMatching } from '@codemirror/language';
import { closeBrackets } from '@codemirror/autocomplete';
import { MergeView } from '@codemirror/merge';

const themeCompartment = new Compartment();
const readOnlyCompartment = new Compartment();
const lintCompartment = new Compartment();
const bracketMatchingCompartment = new Compartment();
const closeBracketsCompartment = new Compartment();

class JsonEditorLsp extends LitElement {
    static properties = {
        document: { type: String },
        isDarkTheme: { type: Boolean },
        isReadOnly: { type: Boolean },
        hasLineNumbers: { type: Boolean },
        isLint: { type: Boolean },
        isMergeMode: { type: Boolean },
        mergeRightContent: { type: String },
        matchBrackets: { type: Boolean },
        autoCloseBrackets: { type: Boolean },
        highlightChanges: { type: Boolean },
        collapseUnchanged: { type: Boolean }
    };

    static styles = css`
        :host {
            display: block;
            height: 100%;
            width: 100%;
        }
        #editor-container {
            height: 100%;
            width: 100%;
            border: 1px solid var(--lumo-contrast-20pct, #ccc);
            border-radius: var(--lumo-border-radius-m, 4px);
            overflow: hidden;
        }
        .cm-editor {
            height: 100%;
        }
        /* Light theme background - applied via host attribute */
        :host([data-theme="light"]) .cm-editor {
            background-color: #ffffff !important;
            color: #000000 !important;
        }
        :host([data-theme="light"]) .cm-gutters {
            background-color: #f5f5f5 !important;
            color: #333333 !important;
            border-right: 1px solid #ddd !important;
        }
        :host([data-theme="light"]) .cm-activeLineGutter {
            background-color: #e0e0e0 !important;
        }
        :host([data-theme="light"]) .cm-activeLine {
            background-color: rgba(0, 0, 0, 0.05) !important;
        }
        :host([data-theme="light"]) .cm-content {
            caret-color: #000000 !important;
        }

        /* Read-only visual indicator */
        :host([data-readonly="true"]) .cm-editor {
            opacity: 0.7;
        }
        :host([data-readonly="true"]) .cm-content {
            cursor: not-allowed;
        }
        :host([data-readonly="true"][data-theme="light"]) .cm-editor {
            background-color: #f8f8f8 !important;
        }
        :host([data-readonly="true"][data-theme="dark"]) .cm-editor {
            background-color: #1e1e1e !important;
        }
        :host([data-readonly="true"]) .cm-gutters {
            opacity: 0.6;
        }
        :host([data-readonly="true"]) #editor-container {
            border-color: var(--lumo-contrast-10pct, #ddd);
        }

        .cm-scroller {
            overflow: auto;
        }

        /* CM6 MergeView styles */
        .cm-mergeView {
            height: 100% !important;
            width: 100% !important;
            display: block !important;
        }
        .cm-mergeViewEditors {
            height: 100% !important;
            width: 100% !important;
            display: flex !important;
            flex-direction: row !important;
        }
        .cm-mergeViewEditor {
            flex: 1 1 0 !important;
            min-width: 0 !important;
            height: 100% !important;
            overflow: hidden !important;
        }
        .cm-mergeViewEditor .cm-editor {
            height: 100% !important;
            width: 100% !important;
        }

        /* Revert controls container - the gap between editors */
        .cm-mergeViewEditors > .cm-merge-revert {
            flex: 0 0 40px !important;
            width: 40px !important;
            min-width: 40px !important;
            position: relative !important;
            background: var(--lumo-contrast-5pct, #f5f5f5) !important;
            border-left: 1px solid var(--lumo-contrast-20pct, #ccc) !important;
            border-right: 1px solid var(--lumo-contrast-20pct, #ccc) !important;
            overflow: visible !important;
        }

        /* Revert buttons inside the gap */
        .cm-mergeViewEditors > .cm-merge-revert > button {
            position: absolute !important;
            left: 50% !important;
            transform: translateX(-50%) !important;
            background: var(--lumo-primary-color, #1976d2) !important;
            color: white !important;
            border: none !important;
            border-radius: 3px !important;
            cursor: pointer !important;
            padding: 2px 8px !important;
            font-size: 14px !important;
            line-height: 1.2 !important;
            z-index: 10 !important;
        }
        .cm-mergeViewEditors > .cm-merge-revert > button:hover {
            background: var(--lumo-primary-color-50pct, #1565c0) !important;
        }

        /* Change gutter - shows markers for changed lines */
        .cm-changeGutter {
            width: 6px !important;
            min-width: 6px !important;
        }
        .cm-changedLineGutter {
            width: 6px !important;
            background: var(--lumo-primary-color, #1976d2) !important;
        }

        /* Spacer widgets for aligning editors at diff chunks */
        .cm-mergeSpacer {
            background: var(--lumo-contrast-5pct, #f5f5f5) !important;
        }

        /* Collapsed unchanged sections */
        .cm-collapsedLines {
            background: var(--lumo-contrast-5pct, #f5f5f5) !important;
            color: var(--lumo-contrast-50pct, #888) !important;
            cursor: pointer;
            padding: 2px 8px;
        }
        .cm-collapsedLines:hover {
            background: var(--lumo-contrast-10pct, #eee) !important;
        }

        /* Change highlighting */
        .cm-changedLine {
            background-color: rgba(255, 200, 0, 0.15) !important;
        }
        .cm-deletedLine {
            background-color: rgba(255, 100, 100, 0.2) !important;
        }
        .cm-insertedLine {
            background-color: rgba(100, 255, 100, 0.2) !important;
        }
        .cm-changedText {
            background-color: rgba(255, 200, 0, 0.35) !important;
        }

        /* Cursor visibility - use theme-appropriate colors */
        .cm-cursor,
        .cm-cursor-primary {
            border-left-width: 2px !important;
            border-left-style: solid !important;
            visibility: visible !important;
            display: block !important;
        }
        /* Dark theme cursor - bright cyan */
        :host([data-theme="dark"]) .cm-cursor,
        :host([data-theme="dark"]) .cm-cursor-primary {
            border-left-color: #00ffff !important;
        }
        /* Light theme cursor - dark blue */
        :host([data-theme="light"]) .cm-cursor,
        :host([data-theme="light"]) .cm-cursor-primary {
            border-left-color: #0000ff !important;
        }
        /* Ensure cursor layer is above other content */
        .cm-cursorLayer {
            z-index: 200 !important;
            pointer-events: none !important;
        }
        /* Merge view specific cursor fixes */
        .cm-mergeViewEditor .cm-cursorLayer {
            z-index: 200 !important;
            overflow: visible !important;
        }
        .cm-mergeViewEditor .cm-cursor,
        .cm-mergeViewEditor .cm-cursor-primary {
            visibility: visible !important;
            display: block !important;
        }
    `;

    constructor() {
        super();
        this.document = '';
        this.isDarkTheme = false;
        this.isReadOnly = false;
        this.hasLineNumbers = true;
        this.isLint = true;
        this.isMergeMode = false;
        this.mergeRightContent = '';
        this.matchBrackets = true;
        this.autoCloseBrackets = true;
        this.highlightChanges = true;
        this.collapseUnchanged = false;

        // Internal state
        this._editor = null;
        this._rightEditor = null;
        this._mergeView = null;
        this._isInternalUpdate = false;

        // CM6 MergeView options (real API)
        this._mergeOptions = {
            revertControls: 'b-to-a',      // 'a-to-b' | 'b-to-a' | null
            orientation: 'a-b',             // 'a-b' | 'b-a'
            gutter: true,                   // show change markers in gutter
            collapseUnchangedMargin: 3,     // lines to show around changes
            collapseUnchangedMinSize: 4     // minimum lines to collapse
        };

        // Scroll sync state
        this._scrollSyncEnabled = false;
        this._scrollSyncing = false;

        // Navigation state for diff chunks
        this._currentChunkIndex = -1;
    }

    render() {
        return html`<div id="editor-container"></div>`;
    }

    firstUpdated() {
        // Set initial attributes for CSS selectors
        this.setAttribute('data-theme', this.isDarkTheme ? 'dark' : 'light');
        this.setAttribute('data-readonly', this.isReadOnly ? 'true' : 'false');
        this._initEditor();
    }

    disconnectedCallback() {
        super.disconnectedCallback();
        this._destroyEditors();
    }

    _destroyEditors() {
        if (this._mergeView) {
            this._mergeView.destroy();
            this._mergeView = null;
            this._editor = null;
            this._rightEditor = null;
        } else {
            if (this._editor) {
                this._editor.destroy();
                this._editor = null;
            }
            if (this._rightEditor) {
                this._rightEditor.destroy();
                this._rightEditor = null;
            }
        }
    }

    _getBaseExtensions(withLinting = true) {
        const extensions = [
            basicSetup,
            keymap.of([indentWithTab]),
            json(),
            themeCompartment.of(this.isDarkTheme ? oneDark : []),
            readOnlyCompartment.of(EditorState.readOnly.of(this.isReadOnly)),
            bracketMatchingCompartment.of(this.matchBrackets ? bracketMatching() : []),
            closeBracketsCompartment.of(this.autoCloseBrackets ? closeBrackets() : [])
        ];

        if (withLinting && this.isLint) {
            extensions.push(lintCompartment.of([linter(jsonParseLinter()), lintGutter()]));
        } else {
            extensions.push(lintCompartment.of([]));
        }

        return extensions;
    }

    _initEditor() {
        const container = this.shadowRoot.getElementById('editor-container');
        if (!container) return;

        // Clear container
        container.innerHTML = '';

        if (this.isMergeMode) {
            this._initMergeView(container);
        } else {
            this._initSingleEditor(container);
        }
    }

    _initSingleEditor(container) {
        const extensions = [
            ...this._getBaseExtensions(true),
            EditorView.updateListener.of(update => {
                if (update.docChanged && !this._isInternalUpdate) {
                    this._onDocumentChanged(update.state.doc.toString());
                }
            })
        ];

        const state = EditorState.create({
            doc: this.document || '',
            extensions
        });

        this._editor = new EditorView({
            state,
            parent: container
        });
    }

    _initMergeView(container) {
        try {
            // Cursor visibility theme - ensures cursor is always visible in merge mode
            const cursorTheme = EditorView.theme({
                ".cm-cursor, .cm-cursor-primary": {
                    display: "block !important",
                    visibility: "visible !important",
                    borderLeftWidth: "2px",
                    borderLeftStyle: "solid",
                    borderLeftColor: this.isDarkTheme ? "#00ffff" : "#0000ff"
                },
                ".cm-cursorLayer": {
                    zIndex: "200 !important"
                }
            });

            // Base extensions for both editors (without read-only, added separately)
            const baseExtensions = [
                basicSetup,
                keymap.of([indentWithTab]),
                json(),
                themeCompartment.of(this.isDarkTheme ? oneDark : []),
                bracketMatchingCompartment.of(this.matchBrackets ? bracketMatching() : []),
                closeBracketsCompartment.of(this.autoCloseBrackets ? closeBrackets() : []),
                cursorTheme
            ];

            const lintExtensions = this.isLint ? [linter(jsonParseLinter()), lintGutter()] : [];

            // Build collapseUnchanged config if enabled
            const collapseConfig = this.collapseUnchanged ? {
                margin: this._mergeOptions.collapseUnchangedMargin,
                minSize: this._mergeOptions.collapseUnchangedMinSize
            } : undefined;

            // Create CM6 MergeView with real API options
            this._mergeView = new MergeView({
                a: {
                    doc: this.document || '',
                    extensions: [
                        ...baseExtensions,
                        ...lintExtensions,
                        readOnlyCompartment.of(EditorState.readOnly.of(this.isReadOnly)),
                        EditorView.updateListener.of(update => {
                            if (update.docChanged && !this._isInternalUpdate) {
                                this.document = update.state.doc.toString();
                                if (this.$server?.onDocumentChanged) {
                                    this.$server.onDocumentChanged(this.document);
                                }
                            }
                        })
                    ]
                },
                b: {
                    doc: this.mergeRightContent || '',
                    extensions: [
                        ...baseExtensions,
                        ...lintExtensions,
                        EditorState.readOnly.of(true)  // Right side always read-only
                    ]
                },
                parent: container,
                orientation: this._mergeOptions.orientation,
                highlightChanges: this.highlightChanges,
                gutter: this._mergeOptions.gutter,
                revertControls: this._mergeOptions.revertControls,
                collapseUnchanged: collapseConfig
            });

            // Store references to individual editors
            this._editor = this._mergeView.a;
            this._rightEditor = this._mergeView.b;

            // Setup scroll sync if enabled
            if (this._scrollSyncEnabled) {
                this._setupScrollSync();
            }
        } catch (error) {
            console.error('[JSON MergeView] Error initializing:', error);
        }
    }

    _setupScrollSync() {
        if (!this._editor || !this._rightEditor) return;

        const leftScroller = this._editor.scrollDOM;
        const rightScroller = this._rightEditor.scrollDOM;

        const syncScroll = (source, target) => {
            if (this._scrollSyncing) return;
            this._scrollSyncing = true;

            // Calculate scroll ratio
            const sourceMax = source.scrollHeight - source.clientHeight;
            const targetMax = target.scrollHeight - target.clientHeight;

            if (sourceMax > 0 && targetMax > 0) {
                const ratio = source.scrollTop / sourceMax;
                target.scrollTop = ratio * targetMax;
            }

            requestAnimationFrame(() => {
                this._scrollSyncing = false;
            });
        };

        leftScroller.addEventListener('scroll', () => {
            if (this._scrollSyncEnabled) {
                syncScroll(leftScroller, rightScroller);
            }
        });

        rightScroller.addEventListener('scroll', () => {
            if (this._scrollSyncEnabled) {
                syncScroll(rightScroller, leftScroller);
            }
        });
    }

    _onDocumentChanged(newValue) {
        if (this.document === newValue) return;
        this.document = newValue;

        if (this.$server?.onDocumentChanged) {
            this.$server.onDocumentChanged(newValue);
        }
    }

    // Public API (called from Vaadin)

    setEditorDocument(element, doc) {
        if (this.document === doc) return;

        this._isInternalUpdate = true;
        try {
            this.document = doc;

            if (this._editor) {
                this._editor.dispatch({
                    changes: {
                        from: 0,
                        to: this._editor.state.doc.length,
                        insert: doc || ''
                    }
                });
            }
        } finally {
            this._isInternalUpdate = false;
        }
    }

    setDarkTheme(isDark) {
        this.isDarkTheme = isDark;
        this.setAttribute('data-theme', isDark ? 'dark' : 'light');
        const effect = themeCompartment.reconfigure(isDark ? oneDark : []);

        if (this._editor) {
            this._editor.dispatch({ effects: effect });
        }
        if (this._rightEditor) {
            this._rightEditor.dispatch({ effects: effect });
        }
    }

    setReadOnly(readOnly) {
        this.isReadOnly = readOnly;
        this.setAttribute('data-readonly', readOnly ? 'true' : 'false');
        const effect = readOnlyCompartment.reconfigure(EditorState.readOnly.of(readOnly));

        if (this._editor) {
            this._editor.dispatch({ effects: effect });
        }
    }

    setLint(lint) {
        this.isLint = lint;
        const effect = lintCompartment.reconfigure(lint ? [linter(jsonParseLinter()), lintGutter()] : []);

        if (this._editor) {
            this._editor.dispatch({ effects: effect });
        }
        if (this._rightEditor) {
            this._rightEditor.dispatch({ effects: effect });
        }
    }

    setMatchBrackets(enabled) {
        this.matchBrackets = enabled;
        const effect = bracketMatchingCompartment.reconfigure(enabled ? bracketMatching() : []);

        if (this._editor) {
            this._editor.dispatch({ effects: effect });
        }
        if (this._rightEditor) {
            this._rightEditor.dispatch({ effects: effect });
        }
    }

    setAutoCloseBrackets(enabled) {
        this.autoCloseBrackets = enabled;
        const effect = closeBracketsCompartment.reconfigure(enabled ? closeBrackets() : []);

        if (this._editor) {
            this._editor.dispatch({ effects: effect });
        }
        if (this._rightEditor) {
            this._rightEditor.dispatch({ effects: effect });
        }
    }

    setMergeModeEnabled(enabled) {
        if (this.isMergeMode === enabled) return;

        this.isMergeMode = enabled;

        // Save current document before rebuilding
        const currentDoc = this.getDocument();
        this.document = currentDoc;

        // Clear container and rebuild editor
        const container = this.shadowRoot.getElementById('editor-container');
        if (container) {
            this._destroyEditors();
            this._initEditor();
        }
    }

    setMergeRightContent(content) {
        this.mergeRightContent = content || '';

        // CM6 MergeView doesn't support dynamic content updates well,
        // so we need to rebuild it
        if (this.isMergeMode && this._mergeView) {
            this._rebuildMergeView();
        } else if (this._rightEditor) {
            this._isInternalUpdate = true;
            try {
                this._rightEditor.dispatch({
                    changes: {
                        from: 0,
                        to: this._rightEditor.state.doc.length,
                        insert: content || ''
                    }
                });
            } finally {
                this._isInternalUpdate = false;
            }
        }
    }

    // CM6 MergeView API methods

    setRevertControls(direction) {
        // direction: 'a-to-b' | 'b-to-a' | null
        this._mergeOptions.revertControls = direction;
        if (this.isMergeMode) {
            this._rebuildMergeView();
        }
    }

    setGutter(enabled) {
        this._mergeOptions.gutter = enabled;
        if (this.isMergeMode) {
            this._rebuildMergeView();
        }
    }

    setHighlightChanges(enabled) {
        this.highlightChanges = enabled;
        // Requires rebuild - CM6 MergeView doesn't support dynamic reconfiguration
        if (this.isMergeMode) {
            this._rebuildMergeView();
        }
    }

    setCollapseUnchanged(enabled) {
        this.collapseUnchanged = enabled;
        // Requires rebuild
        if (this.isMergeMode) {
            this._rebuildMergeView();
        }
    }

    setCollapseUnchangedMargin(margin) {
        this._mergeOptions.collapseUnchangedMargin = margin;
        if (this.isMergeMode && this.collapseUnchanged) {
            this._rebuildMergeView();
        }
    }

    setSyncScroll(enabled) {
        this._scrollSyncEnabled = enabled;
        // If enabling and merge view exists, setup listeners
        if (enabled && this.isMergeMode && this._editor && this._rightEditor) {
            this._setupScrollSync();
        }
    }

    // Legacy method for compatibility - maps to new API
    setMergeOption(option, value) {
        switch (option) {
            case 'revertControls':
                this.setRevertControls(value);
                break;
            case 'gutter':
                this.setGutter(value);
                break;
            case 'highlightChanges':
                this.setHighlightChanges(value);
                break;
            case 'collapseUnchanged':
                this.setCollapseUnchanged(value);
                break;
            case 'syncScroll':
            case 'connect':
                this.setSyncScroll(!!value);
                break;
            default:
                console.warn(`[JSON Editor] Unknown merge option: ${option}`);
        }
    }

    _rebuildMergeView() {
        const currentDoc = this.getDocument();
        const currentRight = this.getMergeRightContent();
        this.document = currentDoc;
        this.mergeRightContent = currentRight;

        const container = this.shadowRoot.getElementById('editor-container');
        if (container) {
            this._destroyEditors();
            this._initEditor();
        }
    }

    goToNextChange() {
        if (this._mergeView) {
            // Use MergeView's built-in chunk navigation
            const chunks = this._mergeView.chunks;
            if (!chunks || chunks.length === 0) return;

            this._currentChunkIndex++;
            if (this._currentChunkIndex >= chunks.length) {
                this._currentChunkIndex = 0;
            }

            const chunk = chunks[this._currentChunkIndex];
            if (chunk && this._editor) {
                this._editor.dispatch({
                    selection: { anchor: chunk.fromA },
                    scrollIntoView: true
                });
            }
        }
    }

    goToPreviousChange() {
        if (this._mergeView) {
            const chunks = this._mergeView.chunks;
            if (!chunks || chunks.length === 0) return;

            this._currentChunkIndex--;
            if (this._currentChunkIndex < 0) {
                this._currentChunkIndex = chunks.length - 1;
            }

            const chunk = chunks[this._currentChunkIndex];
            if (chunk && this._editor) {
                this._editor.dispatch({
                    selection: { anchor: chunk.fromA },
                    scrollIntoView: true
                });
            }
        }
    }

    getMergeRightContent() {
        if (this._rightEditor) {
            return this._rightEditor.state.doc.toString();
        }
        return this.mergeRightContent || '';
    }

    getDocument() {
        if (this._editor) {
            return this._editor.state.doc.toString();
        }
        return this.document || '';
    }

    onRefreshEditor() {
        // CM6 doesn't need explicit refresh like CM5
    }

    appendText(text) {
        if (!this._editor) return;

        const length = this._editor.state.doc.length;
        this._editor.dispatch({
            changes: { from: length, insert: text + '\n' }
        });
    }

    scrollToBottom() {
        if (!this._editor) return;
        this._editor.scrollDOM.scrollTop = this._editor.scrollDOM.scrollHeight;
    }
}

customElements.define('json-editor-lsp', JsonEditorLsp);

File: sapl-vaadin-editor/src/main/java/io/sapl/vaadin/lsp/JsonEditorLsp.java
/*
 * Copyright (C) 2017-2026 Dominic Heutelbeck (dominic@heutelbeck.com)
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.sapl.vaadin.lsp;

import java.io.Serial;
import java.util.ArrayList;
import java.util.List;

import com.vaadin.flow.component.ClientCallable;
import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.HasSize;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.component.dependency.NpmPackage;
import com.vaadin.flow.dom.Element;
import com.vaadin.flow.dom.ElementConstants;

import io.sapl.api.SaplVersion;
import io.sapl.vaadin.DocumentChangedEvent;
import io.sapl.vaadin.DocumentChangedListener;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;

/**
 * JSON editor using CodeMirror 6 with syntax highlighting, linting, and merge
 * view.
 */
@Slf4j
@Tag("json-editor-lsp")
@JsModule("./json-editor-lsp.js")
@NpmPackage(value = "@codemirror/lang-json", version = "6.0.1")
@NpmPackage(value = "@codemirror/merge", version = "6.8.0")
public class JsonEditorLsp extends Component implements HasSize {

    @Serial
    private static final long serialVersionUID = SaplVersion.VERSION_UID;

    private static final String PROP_IS_DARK_THEME       = "isDarkTheme";
    private static final String PROP_IS_READ_ONLY        = "isReadOnly";
    private static final String PROP_HAS_LINE_NUMBERS    = "hasLineNumbers";
    private static final String PROP_IS_LINT             = "isLint";
    private static final String PROP_IS_MERGE_MODE       = "isMergeMode";
    private static final String PROP_MERGE_RIGHT_CONTENT = "mergeRightContent";
    private static final String PROP_MATCH_BRACKETS      = "matchBrackets";
    private static final String PROP_AUTO_CLOSE_BRACKETS = "autoCloseBrackets";
    private static final String PROP_HIGHLIGHT_CHANGES   = "highlightChanges";
    private static final String PROP_COLLAPSE_UNCHANGED  = "collapseUnchanged";

    @Getter
    private String document;

    private final List<DocumentChangedListener> documentChangedListeners = new ArrayList<>();

    /**
     * Creates a JSON editor with default configuration.
     */
    public JsonEditorLsp() {
        this(new JsonEditorLspConfiguration());
    }

    /**
     * Creates a JSON editor with the specified configuration.
     *
     * @param config the editor configuration
     */
    public JsonEditorLsp(JsonEditorLspConfiguration config) {
        getElement().getStyle().set(ElementConstants.STYLE_WIDTH, "100%");
        getElement().getStyle().set(ElementConstants.STYLE_HEIGHT, "100%");
        applyConfiguration(config);
    }

    private void applyConfiguration(JsonEditorLspConfiguration config) {
        Element element = getElement();
        element.setProperty(PROP_IS_DARK_THEME, config.isDarkTheme());
        element.setProperty(PROP_IS_READ_ONLY, config.isReadOnly());
        element.setProperty(PROP_HAS_LINE_NUMBERS, config.isHasLineNumbers());
        element.setProperty(PROP_IS_LINT, config.isLint());
    }

    /**
     * Sets the document content.
     *
     * @param document the JSON document text
     */
    public void setDocument(String document) {
        this.document = document;
        Element element = getElement();
        element.callJsFunction("setEditorDocument", element, document);
        for (var listener : documentChangedListeners) {
            listener.onDocumentChanged(new DocumentChangedEvent(document, false));
        }
    }

    /**
     * Enables or disables the dark theme.
     *
     * @param isDarkTheme true for dark theme
     */
    public void setDarkTheme(boolean isDarkTheme) {
        getElement().callJsFunction("setDarkTheme", isDarkTheme);
    }

    /**
     * Returns whether dark theme is enabled.
     *
     * @return true if dark theme is enabled
     */
    public boolean isDarkTheme() {
        return getElement().getProperty(PROP_IS_DARK_THEME, false);
    }

    /**
     * Enables or disables read-only mode.
     *
     * @param isReadOnly true for read-only mode
     */
    public void setReadOnly(boolean isReadOnly) {
        getElement().callJsFunction("setReadOnly", isReadOnly);
    }

    /**
     * Returns whether the editor is read-only.
     *
     * @return true if read-only
     */
    public boolean isReadOnly() {
        return getElement().getProperty(PROP_IS_READ_ONLY, false);
    }

    /**
     * Enables or disables JSON linting.
     *
     * @param isLint true to enable linting
     */
    public void setLint(boolean isLint) {
        getElement().callJsFunction("setLint", isLint);
    }

    /**
     * Returns whether linting is enabled.
     *
     * @return true if linting is enabled
     */
    public boolean isLint() {
        return getElement().getProperty(PROP_IS_LINT, true);
    }

    /**
     * Enables or disables merge mode (side-by-side diff view).
     *
     * @param enabled true to enable merge mode
     */
    public void setMergeModeEnabled(boolean enabled) {
        getElement().callJsFunction("setMergeModeEnabled", enabled);
    }

    /**
     * Returns whether merge mode is enabled.
     *
     * @return true if merge mode is enabled
     */
    public boolean isMergeModeEnabled() {
        return getElement().getProperty(PROP_IS_MERGE_MODE, false);
    }

    /**
     * Sets the content for the right side of the merge view.
     *
     * @param content the content for the right pane
     */
    public void setMergeRightContent(String content) {
        getElement().callJsFunction("setMergeRightContent", content);
    }

    /**
     * Gets the content from the right side of the merge view.
     *
     * @return the right pane content
     */
    public String getMergeRightContent() {
        return getElement().getProperty(PROP_MERGE_RIGHT_CONTENT, "");
    }

    /**
     * Enables or disables bracket matching highlighting.
     *
     * @param enabled true to enable bracket matching
     */
    public void setMatchBrackets(boolean enabled) {
        getElement().callJsFunction("setMatchBrackets", enabled);
    }

    /**
     * Returns whether bracket matching is enabled.
     *
     * @return true if bracket matching is enabled
     */
    public boolean isMatchBrackets() {
        return getElement().getProperty(PROP_MATCH_BRACKETS, true);
    }

    /**
     * Enables or disables auto-closing of brackets.
     *
     * @param enabled true to enable auto-close brackets
     */
    public void setAutoCloseBrackets(boolean enabled) {
        getElement().callJsFunction("setAutoCloseBrackets", enabled);
    }

    /**
     * Returns whether auto-close brackets is enabled.
     *
     * @return true if auto-close brackets is enabled
     */
    public boolean isAutoCloseBrackets() {
        return getElement().getProperty(PROP_AUTO_CLOSE_BRACKETS, true);
    }

    /**
     * Enables or disables highlighting of changes in merge mode.
     *
     * @param enabled true to highlight changes
     */
    public void setHighlightChanges(boolean enabled) {
        getElement().callJsFunction("setHighlightChanges", enabled);
    }

    /**
     * Returns whether change highlighting is enabled in merge mode.
     *
     * @return true if change highlighting is enabled
     */
    public boolean isHighlightChanges() {
        return getElement().getProperty(PROP_HIGHLIGHT_CHANGES, true);
    }

    /**
     * Enables or disables collapsing of unchanged regions in merge mode.
     *
     * @param enabled true to collapse unchanged regions
     */
    public void setCollapseUnchanged(boolean enabled) {
        getElement().callJsFunction("setCollapseUnchanged", enabled);
    }

    /**
     * Returns whether unchanged region collapsing is enabled in merge mode.
     *
     * @return true if collapsing is enabled
     */
    public boolean isCollapseUnchanged() {
        return getElement().getProperty(PROP_COLLAPSE_UNCHANGED, false);
    }

    /**
     * Enables or disables synchronized scrolling between the two editors in merge
     * mode.
     * When enabled, scrolling one editor will scroll the other proportionally.
     *
     * @param enabled true to enable synchronized scrolling
     */
    public void setSyncScroll(boolean enabled) {
        getElement().callJsFunction("setSyncScroll", enabled);
    }

    /**
     * Enables or disables the change gutter in merge mode.
     * The gutter shows markers next to changed lines.
     *
     * @param enabled true to show the change gutter
     */
    public void setGutter(boolean enabled) {
        getElement().callJsFunction("setGutter", enabled);
    }

    /**
     * Sets the number of context lines to show around changes when collapsing
     * unchanged regions.
     *
     * @param margin number of lines to show (default: 3)
     */
    public void setCollapseUnchangedMargin(int margin) {
        getElement().callJsFunction("setCollapseUnchangedMargin", margin);
    }

    /**
     * Sets a merge view option with a string value.
     * Supported options: "revertControls", "syncScroll"
     *
     * @param option the option name
     * @param value the string value, or null to unset
     */
    public void setMergeOption(String option, String value) {
        getElement().callJsFunction("setMergeOption", option, value);
    }

    /**
     * Sets a merge view option with a boolean value.
     * Supported options: "gutter", "highlightChanges", "collapseUnchanged",
     * "syncScroll"
     *
     * @param option the option name
     * @param value the boolean value
     */
    public void setMergeOption(String option, boolean value) {
        getElement().callJsFunction("setMergeOption", option, value);
    }

    /**
     * Navigates to the next difference in merge mode.
     */
    public void goToNextChange() {
        getElement().callJsFunction("goToNextChange");
    }

    /**
     * Navigates to the previous difference in merge mode.
     */
    public void goToPreviousChange() {
        getElement().callJsFunction("goToPreviousChange");
    }

    /**
     * Scrolls the editor to the bottom.
     */
    public void scrollToBottom() {
        getElement().callJsFunction("scrollToBottom");
    }

    /**
     * Refreshes the editor.
     */
    public void refresh() {
        getElement().callJsFunction("onRefreshEditor");
    }

    /**
     * Appends text to the end of the editor contents.
     *
     * @param text text to append
     */
    public void appendText(String text) {
        getElement().callJsFunction("appendText", text);
    }

    /**
     * Registers a document changed listener.
     *
     * @param listener the listener
     */
    public void addDocumentChangedListener(DocumentChangedListener listener) {
        documentChangedListeners.add(listener);
    }

    /**
     * Removes a document changed listener.
     *
     * @param listener the listener to remove
     */
    public void removeDocumentChangedListener(DocumentChangedListener listener) {
        documentChangedListeners.remove(listener);
    }

    /**
     * Called from JavaScript when the document changes.
     *
     * @param newValue the new document content
     */
    @ClientCallable
    protected void onDocumentChanged(String newValue) {
        document = newValue;
        for (var listener : documentChangedListeners) {
            listener.onDocumentChanged(new DocumentChangedEvent(newValue, true));
        }
    }
}
